### PR TITLE
Annotate how key image exporting/importing works

### PR DESCRIPTION
**Pull request by Doy-lee**
Was re-reading how this works and added some notes on how key images are laid out in memory. Some minor helper variables added to make it easier to understand as well

```
 // NOTE: Exported Key Image File
// [(magic bytes) (ciphertext..................................................................................) (hashed ciphertext signature)]
// [              ((transfer array offset*) (spend public key) (view public key) {(key image) (signature), ...})                              ]
// *The offset in the wallet's transfers this exported key image file contains.
```

* Remove some ambigious overloads for clarity and make it less annoying to
understand.

### CODE REVIEW

File: src/wallet/wallet2.cpp [hunk 1]
```diff
@@ -12978,17 +12978,21 @@ crypto::public_key wallet2::get_tx_pub_key_from_received_outs(const tools::walle
...
-
   std::string data;
-  data.reserve(4 + ski.second.size() * (sizeof(crypto::key_image) + sizeof(crypto::signature)) + 2 * sizeof(crypto::public_key));
-  data.resize(4);
+  const uint32_t offset = ski.first;
+  data.reserve(sizeof(offset) + ski.second.size() * (sizeof(crypto::key_image) + sizeof(crypto::signature)) + 2 * sizeof(crypto::public_key));
+  data.resize(sizeof(offset));
   data[0] = offset & 0xff;
   data[1] = (offset >> 8) & 0xff;
   data[2] = (offset >> 16) & 0xff;
```
Comment (jagerman) [inline src/wallet/wallet2.cpp:12998]: Why is there no `data[3]`?

File: src/wallet/wallet2.cpp [hunk 2]
```diff
@@ -13094,29 +13098,37 @@ uint64_t wallet2::import_key_images(const std::string &filename, uint64_t &spent
...
-  const cryptonote::account_public_address &keys = get_account().get_keys().m_account_address;
-  if (public_spend_key != keys.m_spend_public_key || public_view_key != keys.m_view_public_key)
+  THROW_WALLET_EXCEPTION_IF(offset > m_transfers.size(), error::wallet_internal_error, "Offset larger than known outputs");
+
+  // Validate embedded spend/view public keys
   {
-    THROW_WALLET_EXCEPTION(error::wallet_internal_error, std::string( "Key images from ") + filename + " are for a different account");
+    crypto::public_key public_spend_key = {};
+    crypto::public_key public_view_key  = {};
+    memcpy(&public_spend_key, &data[sizeof(offset)], sizeof(public_spend_key));
```
[the remark this answers is not in the document]
Comment (Doy-lee) [reply src/wallet/wallet2.cpp]: Prefixed with std

File: src/wallet/wallet2.cpp [hunk 3]
```diff
@@ -13094,29 +13098,37 @@ uint64_t wallet2::import_key_images(const std::string &filename, uint64_t &spent
...
-  const crypto::public_key &public_view_key = *(const crypto::public_key*)&data[4 + sizeof(crypto::public_key)];
-  const cryptonote::account_public_address &keys = get_account().get_keys().m_account_address;
-  if (public_spend_key != keys.m_spend_public_key || public_view_key != keys.m_view_public_key)
+  THROW_WALLET_EXCEPTION_IF(offset > m_transfers.size(), error::wallet_internal_error, "Offset larger than known outputs");
+
+  // Validate embedded spend/view public keys
   {
-    THROW_WALLET_EXCEPTION(error::wallet_internal_error, std::string( "Key images from ") + filename + " are for a different account");
+    crypto::public_key public_spend_key = {};
+    crypto::public_key public_view_key  = {};
```
[the remark this answers is not in the document]
Comment (Doy-lee) [reply src/wallet/wallet2.cpp]: Removed

File: src/wallet/wallet2.cpp (hunk 1 above)
Comment (Doy-lee) [reply src/wallet/wallet2.cpp]: Changed to std::memcpy, data[3] is there, seems like github ui goofed up

**Pull request merged**